src/python/yapykaldi/asr.py

```diff
@@ -0,0 +1,161 @@
+        stream = self._p.open(format=self.format, channels=self.channels, rate=self.rate, input=True,
+                              frames_per_buffer=self.chunk)
+
+        logging.info("* start listening")
+
+        frames = []
+        while not self._finalize:
+            data = stream.read(self.chunk)
+            self._queue.put(data)
+            frames.append(data)
+
+        logging.info("* stop listening")
+
+        stream.stop_stream()
+        stream.close()
+        self._p.terminate()
+
+        dt_now = datetime.datetime.now()
+        e_sec = int((dt_now - dt_now.replace(hour=0, minute=0, second=0, microsecond=0)).total_seconds() * 1000000)
+        wav_fname = Template(self.wav_out_fmt).safe_substitute(date=dt_now.strftime('%Y-%m-%d'), time=e_sec)
+        _, ext = os.path.splitext(wav_fname)
+        if not ext == '.wav':
+            wav_fname += '.wav'
+
+        wav_out_path = os.path.join(self.output_dir, wav_fname)
+        if os.path.exists(wav_out_path):
+            raise FileExistsError("Cannot create a new file: {}".format(wav_out_path))
+
+        logging.info("* writing data to '{}'".format(wav_out_path))
+        wav_out = wave.open(wav_out_path, 'wb')
+        wav_out.setnchannels(self.channels)
+        wav_out.setsampwidth(self._p.get_sample_size(self.format))
+        wav_out.setframerate(self.rate)
+        wav_out.writeframes(b''.join(frames))
+        wav_out.close()
+
+    def recognize(self):
+        """Method to start the recognition process on audio stream added to process queue"""
+
+        if self._finalize and (not self._queue):
+            raise Exception("Asr object not initialized for recognition")
+        self.model = KaldiNNet3OnlineModel(self.model_dir)
+        self.decoder = KaldiNNet3OnlineDecoder(self.model)
+
+        while not self._finalize:
+            try:
+                data = self._queue.get(block=True, timeout=self.timeout)
+                data = struct.unpack_from('<%dh' % self.chunk, data)
+            except Exception:
+                break
+            else:
+                logging.info("Recognizing chunk")
+                if self.decoder.decode(self.rate, np.array(data, dtype=np.float32), self._finalize):
+                    decoded_string, _ = self.decoder.get_decoded_string()
+                    logging.info("** {}".format(decoded_string))
+                else:
+                    raise RuntimeError("Decoding failed")
+
+    def interrupt_handle(self, sig, frame):
+        """Interrupt handler that sets the flag to stop recognition and close audio stream"""
+
+        logging.info("Handling interrupt")
+        self._finalize = True
+        time.sleep(3)
+
+    def start(self):
+        logging.info("Starting live speech recognition")
+        # Reset internal states at the start of a new call
+        self._queue = multiprocessing.Queue()
+        self._finalize = False
+
+        process = multiprocessing.Process(None, self.recognize, args=())
```

Comment:
I don't think you actually want a different process but rather a different thread if latency is the thing multiprocessing involves serializing objects go go to a dfifferent process but that is not needed for a different *thead*, IIRC.

Reply:
Multiprocessing is needed because of concurrency. Threads execute one at a time so reading of the audio buffer will be an issue then. Also, latency is not the issue in multiprocessing.

It has occurred due to the object oriented design. The thing that is actually required is to a manager for the shared variables, make recognize and listen into two separate process by inheriting from a multiprocessing base class. Instantiate and close all of them in the start method.